Create rule S7144: Bitbucket keys should not be disclosed (APPSEC-1864) (#4476)
SonarSource/rspec: rules/S7144/secrets/rule.adoc

include::../../../shared_content/secrets/description.adoc[]

== Why is this an issue?

include::../../../shared_content/secrets/rationale.adoc[]

> If attackers gain access to a BitBucket key, they will be able to interact with BitBucket API on behalf of the compromised account.

Reviewer: ```suggestion
If attackers gain access to a BitBucket OAuth credentials, they will be able to interact with BitBucket API on behalf of the compromised account.
```

=== What is the potential impact?

Bitbucket keys allow external services to access resources on a user’s behalf with the OAuth 2.0 protocol.
They are used to authenticate applications, services, or APIs using BitBucket for login and access-control.

Below are some real-world scenarios that illustrate some impacts of an attacker
exploiting the secret.

// Set value that can be used to refer to the type of secret in, for example:
// "An attacker can use this {secret_type} to ..."

> :secret_type: secret

Reviewer: I don't think this is used in the next section that are included.

> include::../../../shared_content/secrets/impact/infrastructure_takeover.adoc[]
> 
> include::../../../shared_content/secrets/impact/data_compromise.adoc[]
> 
> include::../../../shared_content/secrets/impact/data_modification.adoc[]

Reviewer: I think these two are more relevant:
```suggestion
include::../../../shared_content/secrets/impact/source_code_compromise.adoc[]

include::../../../shared_content/secrets/impact/supply_chain_attack.adoc[]
```

== How to fix it

> include::../../../shared_content/secrets/fix/revoke.adoc[]
> 
> include::../../../shared_content/secrets/fix/vault.adoc[]

Reviewer: ```suggestion
include::../../../shared_content/secrets/fix/revoke.adoc[]

include::../../../shared_content/secrets/fix/recent_use.adoc[]

include::../../../shared_content/secrets/fix/vault.adoc[]
```

=== Code examples

:example_secret: MP76PZGLQmw63rxZYJ
:example_name: bitbucket-key

> :example_env: BITBUCKET_KEY

Reviewer: Here I suggest using a custom code example that showcases both OAuth credentials (Key + Secret).
Something like
```
props.set("bitbucket.oauth-key", "MP76PZGLQmw63rxZYJ") // Noncompliant 
props.set("bitbucket.oauth-secret", "ASgzPac3EPbXHbTSw6DyUagJZ8ThnUFG") // Noncompliant
```

include::../../../shared_content/secrets/examples.adoc[]

//=== How does this work?

//=== Pitfalls

//=== Going the extra mile

== Resources

> 

Reviewer: You could add a "Documentation" section add add this link as a reference:
https://support.atlassian.com/bitbucket-cloud/docs/use-oauth-on-bitbucket-cloud/

include::../../../shared_content/secrets/description.adoc[]

== Why is this an issue?

include::../../../shared_content/secrets/rationale.adoc[]

If attackers gain access to a BitBucket OAuth credentials, they will be able to interact with BitBucket API on behalf of the compromised account.

=== What is the potential impact?

Bitbucket keys allow external services to access resources on a user’s behalf with the OAuth 2.0 protocol.
They are used to authenticate applications, services, or APIs using BitBucket for login and access-control.

Below are some real-world scenarios that illustrate some impacts of an attacker
exploiting the secret.

include::../../../shared_content/secrets/impact/source_code_compromise.adoc[]

include::../../../shared_content/secrets/impact/supply_chain_attack.adoc[]

== How to fix it

include::../../../shared_content/secrets/fix/revoke.adoc[]

include::../../../shared_content/secrets/fix/recent_use.adoc[]

include::../../../shared_content/secrets/fix/vault.adoc[]

=== Code examples

> :example_secret: MP76PZGLQmw63rxZYJ
> :example_name: bitbucket-key
> :example_env: BITBUCKET_KEY

Reviewer: ```suggestion
```
